frontend/streamlit_app.py

```diff
@@ -0,0 +1,159 @@
+import streamlit as st
+import requests
+import time
+import datetime
+import pytz
+import os
+import ast
+from PIL import Image
+from utils.feedback import show_feedback_form
+from dotenv import load_dotenv
+from typing import Tuple
+
+def response_generator(user_input: str) -> Tuple[str, str]:
+    """
+    Use the chat endpoint to generate a response to the user's query.
+
+    Args:
+    - user_input (str): The query entered by the user.
+
+    Returns:
+    - tuple: Contains the AI response and sources.
+    """
+    url = os.getenv("CHAT_ENDPOINT", "http://localhost:8000/chatApp")
+
+    headers = {"accept": "application/json", "Content-Type": "application/json"}
+
+    payload = {"query": user_input, "listSources": True, "listContext": True}
+
+    try:
+        response = requests.post(url, headers=headers, json=payload)
+        response.raise_for_status()
+
+        try:
+            data = response.json()
+            if not isinstance(data, dict):
+                st.error("Invalid response format")
+                return None, None
+
+        except ValueError:
+            st.error("Failed to decode JSON response")
+            return None, None
+
+        sources = data.get("sources", "")
+        st.session_state.metadata[user_input] = {
+            "sources": sources,
+            "context": data.get("context", ""),
+        }
+
+        return data.get("response", ""), sources
+
+    except requests.exceptions.RequestException as e:
+        st.error(f"Request failed: {e}")
+        return None, None
+
+def main():
+    load_dotenv()
+
+    img = Image.open("assets/or_logo.png")
+    st.set_page_config(page_title="OR Assistant", page_icon=img)
+
+    deployment_time = datetime.datetime.now(pytz.timezone("UTC"))
+    st.info(f'Deployment time: {deployment_time.strftime("%m/%d/%Y %H:%M")} UTC')
+
+    st.title("OR Assistant")
+
+    if not os.getenv("CHAT_ENDPOINT"):
+        st.warning(
+            "The CHAT_ENDPOINT environment variable is not set or is empty. DEFAULT: http://localhost:8000/chatApp"
+        )
+
+    if "chat_history" not in st.session_state:
+        st.session_state.chat_history = []
+    if "metadata" not in st.session_state:
+        st.session_state.metadata = {}
+
+    if not st.session_state.chat_history:
+        st.session_state.chat_history.append({
+            "content": "Hi, I am the OpenROAD assistant. Type your query about OpenROAD",
+            "role": "ai",
+        })
+
+    for message in st.session_state.chat_history:
+        with st.chat_message(message["role"]):
+            st.markdown(message["content"])
+
+    user_input = st.chat_input("Enter your queries ...")
+
+    if user_input:
+        st.session_state.chat_history.append({"content": user_input, "role": "user"})
+
+        with st.chat_message("user"):
+            st.markdown(user_input)
+
+        response, sources = response_generator(user_input)
```

Comment:
Have we implemented conversational memory? i.e. is the assistant able to recall past questions in the conversation history?

Reply:
Not yet, @palaniappan-r  wanted to do that in the backend.